doc/design/inference.md

```diff
@@ -0,0 +1,178 @@
+- `Place` is used to decide which device the program will run on. There are two types of `Place` in the current framework, `CPUPlace` for CPU and `CUDAPlace` for CUDA GPU.
+- `Scope` in Fluid likes the concept in programming languages. It is an association of a name to variable. Global variables in the same `Scope` should have different names. However, there is no restrictions on names of variables in different local scopes. Users have to specify a `Scope` to run a program.
+- `Executor` can be constructed by a user specified place, and provides a unified way to execute a `ProgramDesc` in a `Scope`.
+
+All the three concepts compose the execution environment, that is `Runtime` for inference.
+
+```c++
+class Runtime {
+ public:
+  Runtime(/* CPU or GPU */);
+
+ private:
+  platform::Place* place;
+  framework::Scope* scope;
+  framework::Executor* executor;
+};
+```
+
+1. A program can run on different `Runtime`s.
+   Users can define a runtime for CPU and another runtime for CUDA GPU, and the inference program can run on the two runtimes at the same time. Or users can define two runtimes for CUDA GPU to run the inference program on different GPU devices.
+1. It is possible to share parameters among different programs.
+   Different program can run on the same `Runtime`, so that parameters with the same name will be shared.
+1. Program running on different threads can share parameters.
+   Multi-threads can be launched to run an inference program in parallel on the same `Runtime`.
+
+## Inference Engine
+
+### Why need an Inference Engine?
+
+With `ProgramBuilder` and `Runtime`, user can write codes to do inference.
+Apart from the concepts introduced specially for inference in this design doc, users need handle the details of feed and fetch data, by calling `framework::SetFeedVariable` and `framework::GetFetchVariable`.
+In addition, users need to run the `startup_program` manually to load parameters for each runtime.
+A simple example is listed as following.
+
+```cpp
+ProgramBuilder builder("mnist.paddle");
+Runtime runtime("CPU");
+
+// Run the startup_program once to load parameters for the specified runtime
+runtime.Executor()->Run(builder.StartupProgram(), runtime.Scope(), 0, true, true);
+
+// Run the main_program many times
+for (...) {
+  framework::LoDTensor input;
+  framework::SetFeedVariable(runtime.Scope(), input, ...);
+  runtime.Executor()->Run(builder.MainProgram(), runtime.Scope(), 0, true, true);
+  framework::LoDTensor output;
+  framework::GetFetchVariable(runtime.Scope(), output, ...);
+}
+```
+
+To simplify the interfaces, we design a new structure, `InferenceEngine`.
+
+### Design of Inference Engine
+
+1. An `InferenceEngine` can be constructed by a `ProgramBuilder`.
+1. An `InferenceEngine` also holds pointer to the current `Runtime`. Users can call `SetRuntime()` to set the current runtime, and the `startup_program` will be run once to initialize parameters for this runtime.
```

Comment:
Change the sequence numbers here ? Replace all 1. to 2. , 3. 4. etc 
holds pointer => holds a pointer

Reply:
It is OK to use `1.` for all list items, and Markdown will translate them to `1. 2.  3. ...` automatically.
Else done.